### PR TITLE
[NP-3231] Scrolling Step Wizard

### PR DESCRIPTION
**Pull request by KernelDeimos**
Branch says "Test" but that's because I was unsure if a Git operation would work.
Screen Shot 2021-03-14 at 11 20 00 PM

### CODE REVIEW

File: src/foam/dao/index/TreeIndex.js [hunk 1]
```diff
@@ -321,6 +321,9 @@ foam.CLASS({
 
   methods: [
     function init() {
+      if ( ! this.root && ! this.index ) {
+        debugger;
```
Comment (kgrgreer) [inline src/foam/dao/index/TreeIndex.js]: stale

File: src/foam/u2/crunch/FlatteningCapabilityWizardlet.js [hunk 2]
```diff
@@ -13,15 +13,63 @@ foam.CLASS({
     Displays prerequisite capabilities in the same wizardlet.
   `,
 
+  requires: [
+    'foam.core.ArraySlot',
```
Comment (kgrgreer) [inline src/foam/u2/crunch/FlatteningCapabilityWizardlet.js]: No trailing commas.

File: src/foam/u2/crunch/wizardflow/AutoSaveWizardletsAgent.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  imports: [
+    'wizardlets'
+  ],
+
+  implements: [
+    'foam.core.ContextAgent'
+  ],
+
+  methods: [
+    async function execute() {
```
Comment (kgrgreer) [inline src/foam/u2/crunch/wizardflow/AutoSaveWizardletsAgent.js:23]: Why is this async?

File: src/foam/u2/crunch/wizardflow/StepWizardAgent.js [hunk 4]
```diff
@@ -18,6 +18,7 @@ foam.CLASS({
     // 'wizardConfig',
     'pushView',
     'popView',
+    'ctrl',
```
Comment (kgrgreer) [inline src/foam/u2/crunch/wizardflow/StepWizardAgent.js:17]: sort imports

File: src/foam/u2/wizard/ScrollingStepWizardView.js [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+foam.CLASS({
+  package: 'foam.u2.wizard',
+  name: 'ScrollingStepWizardView',
+  extends: 'foam.u2.wizard.IncrementalStepWizardView',
+  mixins: ['foam.u2.wizard.WizardletRenderUtils'],
+  documentation: `Displays all wizardlets in a scrolling page.`,
+
+  messages: [
+    { name: 'NO_ACTION_LABEL', message: 'Done' },
```
Comment (kgrgreer) [inline src/foam/u2/wizard/ScrollingStepWizardView.js]: trailing comma

File: src/foam/u2/wizard/ScrollingStepWizardView.js [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+          if ( ! el ) {
+            console.error('missing element', wizardPositionElements[hash]);
+            continue;
+          }
+          if ( test_visible(el) ) {
+            if ( ! minTopPosition || pos.compareTo(minTopPosition) < 0 )
+              minTopPosition = pos;
+          }
+        }
+
```
Comment (kgrgreer) [inline src/foam/u2/wizard/ScrollingStepWizardView.js:98]: remove blank line

File: src/foam/u2/wizard/WizardletSection.js [hunk 7]
```diff
@@ -90,11 +90,11 @@ foam.CLASS({
         this.VerticalDetailView,
         'foam.u2.detail.SectionedDetailView'
       );
-      return this.SectionView.create({
+      return  this.SectionView.create({
```
Comment (kgrgreer) [inline src/foam/u2/wizard/WizardletSection.js]: remove the extra space after return.

File: src/foam/u2/wizard/internal/FObjectRecursionSlot.js [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        debug_updateCalls++;
+        if ( parentRefs.includes(obj) ) {
+          this.cleanup();
+          return;
+        }
+        this.value = { obj: obj, cause: this.obj$ };
+        this.subToProps_(obj);
+      };
+      // this.slot(update);
+      this.obj$.sub(() => { update.call(this, this.obj, this.parentRefs); });
```
Comment (kgrgreer) [inline src/foam/u2/wizard/internal/FObjectRecursionSlot.js]: add onDetach()

File: src/foam/u2/wizard/internal/lib.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
+foam.LIB({
+  name: 'foam.u2.wizard.Slot',
+
+  methods: [
+    {
+      name: 'filter',
+      code: function (from, f) {
+        var s = foam.core.SimpleSlot.create({ value: from.get() });
+        from.sub(() => {
```
Comment (kgrgreer) [inline src/foam/u2/wizard/internal/lib.js]: Can this be cleaned up someway? I mean onDetach() somewhere?

File: src/foam/u2/crunch/FlatteningCapabilityWizardlet.js [hunk 10]
```diff
@@ -68,6 +117,22 @@ foam.CLASS({
     async function load() {
       await Promise.all(this.delegates.map(d => d.load()));
       return await this.wao.load(this);
+    },
+    {
+      name: 'getDataUpdateSub',
+      code: function () {
+        // ???: Replace subs with slots to use ArraySlot here
+        var s = foam.core.FObject.create();
+        this.SUPER().sub(() => {
```
Comment (kgrgreer) [inline src/foam/u2/crunch/FlatteningCapabilityWizardlet.js]: add onDetach()?

File: src/foam/u2/crunch/FlatteningCapabilityWizardlet.js [hunk 11]
```diff
@@ -68,6 +117,22 @@ foam.CLASS({
...
+          s.pub(true);
+        })
+        for ( let wizardlet of this.delegates ) {
+          wizardlet.getDataUpdateSub().sub(() => {
```
Comment (kgrgreer) [inline src/foam/u2/crunch/FlatteningCapabilityWizardlet.js]: add onDetach()?

File: src/foam/u2/crunch/wizardflow/AutoSaveWizardletsAgent.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+      for ( let wizardlet of this.wizardlets ) {
+        wizardlet.getDataUpdateSub().sub(() => {
```
Comment (kgrgreer) [inline src/foam/u2/crunch/wizardflow/AutoSaveWizardletsAgent.js:26]: add onDetach()?

File: src/foam/u2/wizard/ScrollingStepWizardView.js [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+        return data$wizardlets.filter(w => w.submit).length > 0;
+      }
+    }
+  ],
+
+  methods: [
+    function initE() {
+      var self = this;
+      window.testing_ = self;
+      this.scrollWizardPosition$.sub(() => {
```
Comment (kgrgreer) [inline src/foam/u2/wizard/ScrollingStepWizardView.js]: add onDetach()?

File: src/foam/u2/crunch/wizardflow/AutoSaveWizardletsAgent.js (hunk 3 above)
Comment (KernelDeimos) [reply src/foam/u2/crunch/wizardflow/AutoSaveWizardletsAgent.js:23]: This is done to match the interface. `foam.util.async.Sequence` expects this method to return a Promise.

File: src/foam/u2/crunch/wizardflow/AutoSaveWizardletsAgent.js (hunk 12 above)
Comment (KernelDeimos) [reply src/foam/u2/crunch/wizardflow/AutoSaveWizardletsAgent.js:26]: Added TODO for now

File: src/foam/u2/tag/CircleIndicator.js [hunk 14]
```diff
@@ -129,6 +133,15 @@ foam.CLASS({
         .end();
       }
 
+      if ( this.indicateProcessing ) {
+        this.start(this.LoadingSpinner)
+          .style({
+            'width': '' + this.size + 'px',
```
Comment (kgrgreer) [inline src/foam/u2/tag/CircleIndicator.js]: Is the '' + needed?

Comment (KernelDeimos) [reply src/foam/u2/tag/CircleIndicator.js]: Thanks - it appears there was a little bit of witch-doctoring in there

**Pull request merged**